File: flask/parser.py
# -*- coding: utf-8 -*-
"""
Parsowanie CSV-ów telemetrycznych z EdgeTX (i innych radii OpenTX-pochodnych).
Elastyczne wykrywanie kolumn — pozycja kolumny GPS może się różnić u różnych
użytkowników (zależnie od skonfigurowanych sensorów), nazewnictwo również.
"""

import csv
import os
import re


ENCODINGS = ("utf-8-sig", "utf-8", "cp1250", "latin-1")

_PAIR_SEP = re.compile(r"[\s,;]+")


def open_csv(path):
    for enc in ENCODINGS:
        try:
            with open(path, "r", encoding=enc, newline="") as f:
                return f.read()
        except UnicodeDecodeError:
            continue
    with open(path, "r", encoding="utf-8", errors="replace", newline="") as f:
        return f.read()


def _is_valid_pair(lat, lon):
    if lat is None or lon is None:
        return False
    if not (-90.0 <= lat <= 90.0):
        return False
    if not (-180.0 <= lon <= 180.0):
        return False
    if abs(lat) < 1e-9 and abs(lon) < 1e-9:
        return False
    return True


def _try_parse_pair(text):
    if text is None:
        return None
    s = text.strip()
    if not s:
        return None
    parts = [p for p in _PAIR_SEP.split(s) if p]
    if len(parts) != 2:
        return None
    try:
        lat = float(parts[0])
        lon = float(parts[1])
    except ValueError:
        return None
    return (lat, lon)


def _try_parse_float(text):
    if text is None:
        return None
    s = text.strip()
    if not s:
        return None
    try:
        return float(s)
    except ValueError:
        return None


def _name_score_gps(name):
    n = (name or "").lower()
    if not n:
        return 0
    score = 0
    if "gps" in n:
        score += 10
    if "pos" in n and "position" not in n:
        score += 3
    if "coord" in n:
        score += 4
    return score


def _is_lat_name(name):
    n = (name or "").lower()
    if not n:
        return False
    if "long" in n or "lng" in n:
        return False
    if "alt" in n or "late" in n or "later" in n:
        return False
    return "lat" in n


def _is_lon_name(name):
    n = (name or "").lower()
    if not n:
        return False
    if "long" in n or "lng" in n:
        return True
    if "lon" in n and "long" not in n:
        return True
    return False


def _peek_first_value(rows, col):
    for row in rows:
        v = (row.get(col) or "").strip()
        if v:
            return v
    return ""


def _detect_helper(fieldnames, *needles):
    for n in fieldnames or ():
        low = (n or "").lower()
        for needle in needles:
            if needle in low:
                return n
    return None


def _detect_columns(fieldnames, sample_rows):
    info = {"mode": None}
    if not fieldnames:
        return info

    pair_candidates = sorted(
        ((_name_score_gps(n), n) for n in fieldnames),
        key=lambda x: -x[0],
    )
    chosen_pair = None
    for score, name in pair_candidates:
        if score <= 0:
            break
        sample = _peek_first_value(sample_rows, name)
        if sample and _try_parse_pair(sample):
            chosen_pair = name
            break

    lat_col = next((n for n in fieldnames if _is_lat_name(n)), None)
    lon_col = next((n for n in fieldnames if _is_lon_name(n)), None)
    chosen_split = (lat_col, lon_col) if (lat_col and lon_col) else None

    if chosen_pair:
        info["mode"] = "pair"
        info["gps_col"] = chosen_pair
    elif chosen_split:
        info["mode"] = "split"
        info["lat_col"], info["lon_col"] = chosen_split

    info["time_col"] = _detect_helper(fieldnames, "time")
    info["date_col"] = _detect_helper(fieldnames, "date")
    info["alt_col"] = _detect_helper(fieldnames, "galt", "alt(m)", "altitude") or _detect_helper(fieldnames, "alt")
    info["sats_col"] = _detect_helper(fieldnames, "sats", "satellites")
    info["gspd_col"] = _detect_helper(fieldnames, "gspd", "groundspeed", "gs(", "speed")
    info["hdg_col"] = _detect_helper(fieldnames, "hdg", "heading", "course", "yaw")
    return info


def parse_file(path):
    """
    Czyta CSV i zwraca listę punktów GPS:
      [{"time", "date", "lat", "lon", "alt", "sats", "gspd", "hdg"}, ...]
    """
    text = open_csv(path)
    if not text.strip():
        return []

    reader = csv.DictReader(text.splitlines())
    fieldnames = reader.fieldnames or []
    rows = list(reader)
    if not rows:
        return []

    info = _detect_columns(fieldnames, rows[:50])
    if info["mode"] is None:
        return []

    points = []
    for row in rows:
        if info["mode"] == "pair":
            pair = _try_parse_pair(row.get(info["gps_col"]))
            if not pair:
                continue
            lat, lon = pair
        else:
            lat = _try_parse_float(row.get(info["lat_col"]))
            lon = _try_parse_float(row.get(info["lon_col"]))

        if not _is_valid_pair(lat, lon):
            continue

        def _val(col):
            if not col:
                return ""
            return (row.get(col) or "").strip()

        points.append({
            "time": _val(info.get("time_col")),
            "date": _val(info.get("date_col")),
            "lat": lat,
            "lon": lon,
            "alt": _val(info.get("alt_col")),
            "sats": _val(info.get("sats_col")),
            "gspd": _val(info.get("gspd_col")),
            "hdg": _val(info.get("hdg_col")),
        })
    return points


def is_log_file(path):
    name = os.path.basename(path)
    if name.startswith(".~lock."):
        return False
    if name.endswith(".csv#"):
        return False
    return name.lower().endswith(".csv")


def iter_log_paths(folder):
    out = []
    for root, _dirs, files in os.walk(folder):
        for f in files:
            full = os.path.join(root, f)
            if is_log_file(full):
                out.append(full)
    out.sort()
    return out

File: parser.js
// Parser logów EdgeTX (i podobnych OpenTX-pochodnych) działający w przeglądarce.
// Elastyczne wykrywanie kolumny GPS — pozycja kolumny zależy od konfiguracji
// telemetrii u danego pilota, więc szukamy po nazwie a nie po indeksie.
//
// API: parseLogFile(file) -> Promise<Array<Point>>
// Point: { time, date, lat, lon, alt, sats, gspd, hdg }

(function (global) {
  "use strict";

  const ENCODINGS = ["utf-8", "windows-1250", "iso-8859-1"];
  const PAIR_SEP = /[\s,;]+/;

  // ---- Odczyt z fallbackiem kodowania ---------------------------------------

  async function readFileText(file) {
    const buf = await file.arrayBuffer();
    for (const enc of ENCODINGS) {
      try {
        const dec = new TextDecoder(enc, { fatal: true });
        const text = dec.decode(buf);
        // Strip BOM
        return text.charCodeAt(0) === 0xFEFF ? text.slice(1) : text;
      } catch (_) {
        // try next encoding
      }
    }
    // Ostateczność: utf-8 z zamianą błędnych bajtów
    const dec = new TextDecoder("utf-8", { fatal: false });
    const text = dec.decode(buf);
    return text.charCodeAt(0) === 0xFEFF ? text.slice(1) : text;
  }

  // ---- Lekki parser CSV (obsługuje cudzysłowy i przecinki w polach) ---------

  function parseCSVLine(line) {
    const out = [];
    let cur = "";
    let inQ = false;
    for (let i = 0; i < line.length; i++) {
      const c = line[i];
      if (inQ) {
        if (c === '"') {
          if (line[i + 1] === '"') { cur += '"'; i++; }
          else { inQ = false; }
        } else {
          cur += c;
        }
      } else {
        if (c === ",") { out.push(cur); cur = ""; }
        else if (c === '"' && cur.length === 0) { inQ = true; }
        else { cur += c; }
      }
    }
    out.push(cur);
    return out;
  }

  function parseCSV(text) {
    const lines = text.split(/\r?\n/).filter((l) => l.length > 0);
    if (lines.length === 0) return { fields: [], rows: [] };
    const fields = parseCSVLine(lines[0]).map((s) => s.trim());
    const rows = [];
    for (let i = 1; i < lines.length; i++) {
      const cells = parseCSVLine(lines[i]);
      const row = {};
      for (let k = 0; k < fields.length; k++) {
        row[fields[k]] = cells[k] !== undefined ? cells[k] : "";
      }
      rows.push(row);
    }
    return { fields, rows };
  }

  // ---- Walidacja punktu ------------------------------------------------------

  function isValidPair(lat, lon) {
    if (lat == null || lon == null || Number.isNaN(lat) || Number.isNaN(lon)) return false;
    if (lat < -90 || lat > 90) return false;
    if (lon < -180 || lon > 180) return false;
    if (Math.abs(lat) < 1e-9 && Math.abs(lon) < 1e-9) return false; // brak fixa
    return true;
  }

  function tryParsePair(text) {
    if (text == null) return null;
    const s = String(text).trim();
    if (!s) return null;
    const parts = s.split(PAIR_SEP).filter(Boolean);
    if (parts.length !== 2) return null;
    const lat = parseFloat(parts[0]);
    const lon = parseFloat(parts[1]);
    if (Number.isNaN(lat) || Number.isNaN(lon)) return null;
    return [lat, lon];
  }

  function tryParseFloat(text) {
    if (text == null) return null;
    const s = String(text).trim();
    if (!s) return null;
    const v = parseFloat(s);
    return Number.isNaN(v) ? null : v;
  }

  // ---- Detekcja kolumn -------------------------------------------------------

  function nameScoreGps(name) {
    const n = (name || "").toLowerCase();
    if (!n) return 0;
    let s = 0;
    if (n.includes("gps")) s += 10;
    if (n.includes("pos") && !n.includes("position")) s += 3;
    if (n.includes("coord")) s += 4;
    return s;
  }

  function isLatName(name) {
    const n = (name || "").toLowerCase();
    if (!n) return false;
    if (n.includes("long") || n.includes("lng")) return false;
    if (n.includes("alt") || n.includes("late") || n.includes("later")) return false;
    return n.includes("lat");
  }

  function isLonName(name) {
    const n = (name || "").toLowerCase();
    if (!n) return false;
    if (n.includes("long") || n.includes("lng")) return true;
    if (n.includes("lon") && !n.includes("long")) return true;
    return false;
  }

  function peekFirstValue(rows, col) {
    for (const r of rows) {
      const v = (r[col] || "").trim();
      if (v) return v;
    }
    return "";
  }

  function detectHelper(fields, ...needles) {
    for (const f of fields) {
      const low = (f || "").toLowerCase();
      for (const needle of needles) {
        if (low.includes(needle)) return f;
      }
    }
    return null;
  }

  function detectColumns(fields, sample) {
    const info = { mode: null };
    if (!fields || fields.length === 0) return info;

    // Tryb A: pojedyncza kolumna z 'lat lon'
    const candidates = fields
      .map((n) => ({ score: nameScoreGps(n), name: n }))
      .filter((c) => c.score > 0)
      .sort((a, b) => b.score - a.score);

    let chosenPair = null;
    for (const c of candidates) {
      const sampleVal = peekFirstValue(sample, c.name);
      if (sampleVal && tryParsePair(sampleVal)) {
        chosenPair = c.name;
        break;
      }
    }

    // Tryb B: osobne lat/lon
    const latCol = fields.find(isLatName) || null;
    const lonCol = fields.find(isLonName) || null;
    let chosenSplit = null;
    if (latCol && lonCol) {
      chosenSplit = [latCol, lonCol];
    }

    if (chosenPair) {
      info.mode = "pair";
      info.gpsCol = chosenPair;
    } else if (chosenSplit) {
      info.mode = "split";
      info.latCol = chosenSplit[0];
      info.lonCol = chosenSplit[1];
    }

    info.timeCol = detectHelper(fields, "time");
    info.dateCol = detectHelper(fields, "date");
    info.altCol = detectHelper(fields, "galt", "alt(m)", "altitude") || detectHelper(fields, "alt");
    info.satsCol = detectHelper(fields, "sats", "satellites");
    info.gspdCol = detectHelper(fields, "gspd", "groundspeed", "gs(", "speed");
    info.hdgCol = detectHelper(fields, "hdg", "heading", "course", "yaw");
    return info;
  }

  // ---- Główne API ------------------------------------------------------------

  function isLogFile(file) {
    const name = (file.name || "").toLowerCase();
    if (name.startsWith(".~lock.")) return false;
    if (name.endsWith(".csv#")) return false;
    return name.endsWith(".csv");
  }

  async function parseLogFile(file) {
    const text = await readFileText(file);
    if (!text.trim()) return [];

    const { fields, rows } = parseCSV(text);
    if (rows.length === 0) return [];

    const info = detectColumns(fields, rows.slice(0, 50));
    if (info.mode === null) return [];

    const valOf = (row, col) => (col ? (row[col] || "").trim() : "");

    const points = [];
    for (const row of rows) {
      let lat, lon;
      if (info.mode === "pair") {
        const pair = tryParsePair(row[info.gpsCol]);
        if (!pair) continue;
        [lat, lon] = pair;
      } else {
        lat = tryParseFloat(row[info.latCol]);
        lon = tryParseFloat(row[info.lonCol]);
      }
      if (!isValidPair(lat, lon)) continue;

      points.push({
        time: valOf(row, info.timeCol),
        date: valOf(row, info.dateCol),
        lat,
        lon,
        alt: valOf(row, info.altCol),
        sats: valOf(row, info.satsCol),
        gspd: valOf(row, info.gspdCol),
        hdg: valOf(row, info.hdgCol),
      });
    }
    return points;
  }

  global.FpvParser = { parseLogFile, isLogFile };
})(window);
